MoreCyclopsUpgrades/CyclopsUpgrades/UpgradeHandler.cs

```diff
@@ -0,0 +1,139 @@
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if <see cref="Count"/> is at least 1; otherwise, <c>false</c>.
+        /// </value>
+        public bool HasUpgrade => this.Count > 0;
+
+        /// <summary>
+        /// This event is invoked when upgrades are cleared right before being re-counted.
+        /// This happens every time upgrades are changed.
+        /// </summary>
+        public UpgradeEvent OnClearUpgrades;
+
+        /// <summary>
+        /// This event is invoked when a copy of this module's <see cref="techType"/> is found and counted.
+        /// This will happen for each copy found in the cyclops.
+        /// </summary>
+        public UpgradeEventSlotBound OnUpgradeCounted;
+
+        /// <summary>
+        /// This event is invoked after all upgrade modules have been found.
+        /// This instance's event will only be called if there is at least 1 copy of this upgrade module's <see cref="techType"/>.
+        /// </summary>
+        public UpgradeEvent OnFinishedUpgrades;
+
+        /// <summary>
+        /// This event is invoked after all upgrade modules have been found and counted,
+        /// but only if this is the first time that the <see cref="MaxCount"/> of upgrades has been achived.
+        /// </summary>
+        public Action OnFirstTimeMaxCountReached;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="UpgradeHandler"/> class.
+        /// </summary>
+        /// <param name="techType">The TechType of the upgrade module.</param>
+        public UpgradeHandler(TechType techType)
+        {
+            this.techType = techType;
+        }
+
+        internal virtual void UpgradesCleared(SubRoot cyclops)
+        {
+            count = 0;
+            OnClearUpgrades?.Invoke(cyclops);
+        }
+
+        internal virtual void UpgradeCounted(SubRoot cyclops, Equipment modules, string slot)
+        {
+            count++;
+            OnUpgradeCounted?.Invoke(cyclops, modules, slot);
+        }
+
+        internal virtual void UpgradesFinished(SubRoot cyclops)
+        {
+            if (count == 0)
+                return;
+
+            if (count > this.MaxCount)
+                return;
+
+            OnFinishedUpgrades?.Invoke(cyclops);
+
+            if (!maxedOut) // If we haven't maxed out before, check this block
+            {
+                maxedOut = count == this.MaxCount; // Are we maxed out now?
+
+                if (maxedOut) // If we are, invoke the event
+                    OnFirstTimeMaxCountReached?.Invoke();
+            }
+            else // If we are in this block, that means we maxed out in a previous cycle
+            {
+                maxedOut = count == this.MaxCount; // Evaluate this again in case an upgrade was removed
+            }
+        }
+
+        internal virtual void RegisterSelf(IDictionary<TechType, UpgradeHandler> dictionary)
+        {
+            QuickLogger.Debug($"{techType} upgrade registered");
```

Comment:
Let's add an optional string field for setting a debug log name.  
And then promote this log to `Info` level.